src/Features/CSharp/Portable/InternalUtilities/CSharpOperationAnalysisHelpers.cs

```diff
@@ -0,0 +1,941 @@
+                        methodSymbol.Parameters.First().Type.SpecialType == SpecialType.System_Object &&
+                        (GetBlockOfMethodBody(methodBodyOperation)) is IBlockOperation
+                        {
+                            Operations: [IReturnOperation
+                            {
+                                ReturnedValue: IInvocationOperation
+                                {
+                                    Instance: IInstanceReferenceOperation,
+                                    Arguments: [IArgumentOperation { Value: IOperation arg }]
+                                }
+                            }]
+                        } && arg.WalkDownConversion() is IParameterReferenceOperation { Parameter: IParameterSymbol param }
+                        && param.Equals(methodSymbol.Parameters.First()))
+                    {
+                        // in this case where we have an Equals(C? other) from IEquatable but the current one
+                        // is Equals(object? other), we accept something of the form:
+                        // return Equals(other as C);
+                        return true;
+                    }
+
+                    // otherwise we check to see which fields are compared (either by themselves or through properties)
+                    var actualFields = GetEqualizedFields(methodBodyOperation, methodSymbol);
+                    return !actualFields.HasDuplicates(SymbolEqualityComparer.Default) &&
+                            actualFields.SetEquals(expectedComparedFields);
+                }
+            }
+
+            return false;
+        }
+
+        public static bool IsSimpleHashCodeMethod(
+            Compilation compilation,
+            IMethodSymbol methodSymbol,
+            IMethodBodyOperation methodOperation,
+            ImmutableArray<IFieldSymbol> expectedHashedFields)
+        {
+            if (methodSymbol.Name == nameof(GetHashCode) &&
+                methodSymbol.Parameters.IsEmpty &&
+                UseSystemHashCode.Analyzer.TryGetAnalyzer(compilation, out var analyzer))
+            {
+                // Hash Code method, see if it would be a default implementation that we can remove
+                var (_, members, _) = analyzer.GetHashedMembers(
+                    methodSymbol, methodOperation.BlockBody ?? methodOperation.ExpressionBody);
+                if (members != default)
+                {
+                    // the user could access a member using either the property or the underlying field
+                    // so anytime they access a property instead of the underlying field we convert it to the
+                    // corresponding underlying field
+                    var actualMembers = members
+                        .SelectAsArray(UnwrapPropertyToField).WhereNotNull().AsImmutable();
+
+                    return !actualMembers.HasDuplicates(SymbolEqualityComparer.Default) &&
```

Comment:
Why are duplicates bad?

Reply:
I was expecting every field to be hashed exactly once, no more and no less, but I suppose that's not a strict requirement here. I think I was keeping the logic that the primary constructor method used, since we don't want to assign to the same property multiple times. I suppose the equals methods could also allow duplicates (they currently don't)